corehq/motech/repeaters/repeater_generators.py

```diff
@@ -261,6 +266,103 @@ def get_test_payload(self, domain):
         )
 
 
+@attr.s
+class CaseTypeReferralConfig(object):
+    #if false, listed_properties is a whitelist
+    use_blacklist = attr.ib()
+    listed_properties = attr.ib()
+    constant_properties = attr.ib()
+
+
+class ReferCasePayloadGenerator(BasePayloadGenerator):
+
+    def get_payload(self, repeat_record, payload_doc):
+
+        def _update_case_id(original_case_id, case_id_map):
+            if original_case_id in case_id_map:
+                new_case_id = case_id_map[original_case_id]
+            else:
+                new_case_id = uuid4().hex
+                case_id_map[original_case_id] = new_case_id
+            return new_case_id
+
+        def _update_case_properties_with_blacklist(case, config):
+            for name in config.listed_properties:
+                if name in case.case_json:
+                    del case.case_json[name]
+
+        def _update_case_properties_with_whitelist(case, config):
+            new_json = {}
+            for name in config.listed_properties:
+                if name in case.case_json:
+                    new_json[name] = case.case_json[name]
+            case.case_json = new_json
+
+        def _set_constant_properties(case, config):
+            for name, value in config.constant_properties:
+                case.case_json[name] = value
+
+        case_ids_to_forward = payload_doc.get_case_property('cases_to_forward').split(' ')
+        new_owner = payload_doc.get_case_property('new_owner')
+        cases_to_forward = CaseAccessors(payload_doc.domain).get_cases(case_ids_to_forward)
+        case_ids_to_forward = set(case_ids_to_forward)
+        included_case_types = payload_doc.get_case_property('case_types').split(' ')
```

Comment:
Is there value in passing `case_types` in the payload when you could derive it from the case ids?

Reply:
mostly it just ensures we are passing cases of the expected type. if a case of a different type is unintentionally added it will fail the fail the transfer because we wont have a config for that case type. @ctsims was there another reason you wanted this?

Reply:
validation is a good reason